File: 백준/Gold/22866. 탑 보기/탑 보기.java
import java.util.*;
import java.io.*;

public class Main {
	public static void main(String[] args) throws IOException{
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		n = Integer.parseInt(br.readLine());
		arr = new int[n+1];
		st = new StringTokenizer(br.readLine());
		for(int i = 1; i<=n; i++) {
			arr[i] = Integer.parseInt(st.nextToken());
		}
		
		Stack<Building> stk = new Stack<>();
		cnt = new int[n+1];
		near = new int[n+1][2];
		for(int row[] : near) {
			Arrays.fill(row, Integer.MAX_VALUE / 2);
		}
		for(int i = 1; i<=n; i++) {
			while(!stk.isEmpty() && stk.peek().height <= arr[i])
				stk.pop();
			
			cnt[i] += stk.size();
			
			if(!stk.isEmpty()) {
				Building peek = stk.peek();
				int gap = Math.abs(peek.idx - i);
				
				if(gap < near[i][0]) {
					near[i][0] = gap;
					near[i][1] = peek.idx;
				}else if(gap == near[i][0] && peek.idx < near[i][1]) {
					near[i][1] = peek.idx;
				}
			}
			
			stk.push(new Building(i, arr[i]));
		}
		
		stk = new Stack<>();
		
		for(int i = n; i>= 1; i--) {
			while(!stk.isEmpty() && stk.peek().height <= arr[i])
				stk.pop();
			
			cnt[i] += stk.size();
			if(!stk.isEmpty()) {
				Building peek = stk.peek();
				int gap = Math.abs(peek.idx - i);
				
				if(gap < near[i][0]) {
					near[i][0] = gap;
					near[i][1] = peek.idx;
				}else if(gap == near[i][0] && peek.idx < near[i][1]) {
					near[i][1] = peek.idx;
				}
			}
			
			stk.push(new Building(i, arr[i]));
		}
		
		StringBuilder sb = new StringBuilder();
		for(int i = 1; i<=n; i++) {
			if(cnt[i] == 0)
				sb.append(0).append('\n');
			else {
				sb.append(cnt[i]).append(' ').append(near[i][1]).append('\n');
			}
		}
		System.out.println(sb);
		
	}
	static StringTokenizer st;
	static int arr[],n, cnt[], near[][];
	static class Building{
		int idx, height;
		public Building(int idx, int height) {
			this.idx = idx;
			this.height = height; 
		}
	}
}

File: PyPy3/백준/Gold/22866. 탑 보기/탑 보기.py
from collections import namedtuple

if __name__ == "__main__":
    N = int(input())
    arr = [0]+ list(map(int,input().split()))
    Building = namedtuple('Building', ['idx', 'height'])

    stk = []
    INF = float('inf')
    cnt = [0 for _ in range(N + 1)]
    near = [[INF, 0] for _ in range(N+1)]

    for i in range(1, N+1):
        while stk and stk[-1].height <= arr[i]:
            stk.pop()

        cnt[i] += len(stk)

        if stk:
            top = stk[-1]

            gap = abs(top.idx - i)

            if gap < near[i][0]:
                near[i][0] = gap
                near[i][1] = top.idx
            elif gap == near[i][0] and top.idx < near[i][1]:
                near[i][1] = top.idx

        stk.append(Building(i, arr[i]))




    stk = []

    for i in range(N,0,-1):
        while stk and stk[-1].height <= arr[i]:
            stk.pop()

        cnt[i] += len(stk)
        if stk:
            top = stk[-1]

            gap = abs(top.idx - i)

            if gap < near[i][0]:
                near[i][0] = gap
                near[i][1] = top.idx
            elif gap == near[i][0] and top.idx < near[i][1]:
                near[i][1] = top.idx

        stk.append(Building(i, arr[i]))

    ans = []

    for i in range(1,N+1):
        if cnt[i] == 0:
            ans.append(f"{0}")
        else:
            ans.append(f"{cnt[i]} {near[i][1]}")

    print('\n'.join(ans))
